plugins/woocommerce/includes/wc-cart-functions.php

```diff
@@ -188,8 +188,20 @@ function wc_clear_cart_after_payment() {
 		$order = wc_get_order( WC()->session->order_awaiting_payment );
 
 		if ( $order instanceof WC_Order && $order->get_id() > 0 ) {
-			// If the order has not failed, or is not pending, the order must have gone through.
-			if ( ! $order->has_status( array( 'failed', 'pending', 'cancelled' ) ) ) {
+			// If the order status is neither pending, failed, nor cancelled, the order must have gone through.
+			$did_order_succeed = ! $order->has_status( array( 'failed', 'pending', 'cancelled' ) );
+
+			/**
+			 * Determine whether the cart should be cleared after payment.
+			 *
+			 * @since 8.7.1
+			 * @param bool $should_clear_cart_after_payment Whether the cart should be cleared after payment.
+			 * @param WC_Order $order The order.
+			 * @return bool Whether the cart should be cleared after payment.
+			 */
+			$should_clear_cart_after_payment = apply_filters( 'woocommerce_should_clear_cart_after_payment', $did_order_succeed, $order );
```

Comment:
I apologize if I missed earlier discussion about this, but might it be confusing (for other developers) that they can exercise control over cart clearance *here,* but not if execution enters the preceding block (I'm referring to the other `WC->cart->empty_cart()` call a few lines further up)?

Reply:
If we want to make this filter more specific, perhaps we can rename the filter to `woocommerce_clear_cart_after_payment_status_list`.

Reply:
I'm happy to change the filter name if that works for folks.

Reply:
We discussed this offline a bit as well, Barry's suggestion makes sense here that we shouldn't have two different hooks, or a hook that fires some of the time. Can we rework this function a bit and clear cart only once at the end, behind a filter?

Reply:
Thanks for the feedback 🙇  I've tried reworking the function in line with what I think you're suggesting.

Take a look and let me know if it's any closer to what you're looking for.